Adding the Compton Sequence variable during tra file reading

GallegoSav opened the pull request:
I only modified the unbinned part , assuming people will use the cut option on the unbinned file and then save it. 
I added a cut function similar to the energy one.

GallegoSav committed:
Update ReadTraTest.py

GallegoSav committed:
Update UnBinnedData.py

ckarwin commented:
@GallegoSav Is this reedy to be reviewed? If so, please add the necessary coverage and I will check the code.

avalluvan commented on cosipy/data_io/UnBinnedData.py:
```diff
@@ -545,7 +557,7 @@ def write_unbinned_output(self, output_name):
         # Data units:
         units=['keV','s','rad','rad',
-                'rad','rad','rad','rad','cm','deg','deg']
+                'rad','rad','rad','rad','cm','deg','deg','']
```
I don't think any of the subsequent code directly uses the units written to the .fits/.hdf5 file. However, do you think it will be useful to include a more explicit string for the dimensionless quantity (number of compton interactions)? For example, 'number' or 'dimensionless'.

GallegoSav replied:
I remember I had some issues if no units were declared.

avalluvan commented on cosipy/data_io/UnBinnedData.py:
```diff
@@ -718,6 +730,47 @@ def select_data_energy(self, emin, emax, output_name=None, unbinned_data=None):
+        
+        Parameters
+        ----------
+        seqmin :  int
```
Could you explicitly mention that the function returns the range [seqmin, seqmax), i.e., seqmin is included and seqmax is excluded?

avalluvan reviewed (commented):
I think the pull request is complete barring the two minor line comments. It will preserve the number of Compton interactions information which can later be used for event cuts and potentially be propagated to the BinnedData object.

GallegoSav committed:
Merge branch 'cositools:develop' into issue_361

GallegoSav committed:
Update test_unbinned_data_all.py

GallegoSav committed:
Update test_unbinned_data_with_MEGAlib.py

GallegoSav committed:
Update test_unbinned_data_with_MEGAlib.py

GallegoSav committed:
Add files via upload

GallegoSav committed:
Delete cosipy/test_data/unbinned_data_MEGAlib_calc.hdf5

GallegoSav committed:
Add files via upload

GallegoSav committed:
Update ReadTraTest.py

GallegoSav committed:
Update test_unbinned_data_with_MEGAlib.py

GallegoSav committed:
Update UnBinnedData.py

GallegoSav commented:
> @GallegoSav Is this reedy to be reviewed? If so, please add the necessary coverage and I will check the code.

@ckarwin I added the coverage and updated the test scripts. The pull request is ready for your review

ckarwin commented:
@GallegoSav I am reviewing this PR. I tested the dataIO tutorial notebook and it runs ok. 

However, I did find one issue. Take a look at the block of code linked below, where we remove bad photon events: https://github.com/GallegoSav/cosipy/blob/11f64dfe1e00659111bc2dd223caced304c648de/cosipy/data_io/UnBinnedData.py#L206

This removes photons that only have one detector hit. It's pretty rare, but it does happen sometimes. I guess this would correspond to events with CO_seq=1. Since we append all the event parameters event by event before knowing if it's a good event or not, we have to remove the data afterwards for bad events. Essentially, I think you need to add `CO_seq.pop()` under `tt.pop()`.

Another thing to be aware of is that we are planning to move all event selections to an event selection class, utilizing the proper protocol, as part of the work from @israelmcmc. Also note that this event selection is a lower-level selection, and it will require the calculation of a new response, since the same selection needs to be made for the response. Maybe we can merge this PR for now, but then work on moving it to the new protocol in the near future. What do you think @israelmcmc? Do you prefer we start now working with the proper protocol? It's probably not a bad idea. If so, maybe we can just merge the part that saves the CO_seq number to the cosi_dataset, and move the selection method.

ckarwin commented:
@GallegoSav Can you also update your branch so it's synced with develop?

GallegoSav committed:
Merge branch 'cositools:develop' into issue_361

GallegoSav committed:
Update UnBinnedData.py

ckarwin commented:
Ok, thanks for making the change. Regarding the select function, @GallegoSav and I discussed it and decided to keep it in dataIO for now. We can make one new PR to move the select functions from dataIO to the new selection class. This PR can be merged now.

ckarwin reviewed (approved):
Can be merged.
